sycl/include/sycl/ext/oneapi/experimental/group_load_store.hpp

```diff
@@ -0,0 +1,1006 @@
+          typename Properties = decltype(properties())>
+std::enable_if_t<
+    std::is_convertible_v<remove_decoration_t<typename std::iterator_traits<
+                              InputIteratorT>::value_type>,
+                          OutputT> &&
+    detail::is_generic_group_v<Group>>
+group_load(Group g, InputIteratorT in_ptr, sycl::vec<OutputT, N> &out,
+           Properties properties = {}) {
+  group_load(g, in_ptr, span<OutputT, N>(&out[0], N), properties);
+}
+
+// Store API sycl::vec overload.
+template <typename Group, typename InputT, int N, typename OutputIteratorT,
+          typename Properties = decltype(properties())>
+std::enable_if_t<std::is_convertible_v<
+                     InputT, remove_decoration_t<typename std::iterator_traits<
+                                 OutputIteratorT>::value_type>> &&
+                 detail::is_generic_group_v<Group>>
+group_store(Group g, const sycl::vec<InputT, N> &in, OutputIteratorT out_ptr,
+            Properties properties = {}) {
+  group_store(g, span<const InputT, N>(&in[0], N), out_ptr, properties);
+}
+
+// Load API span + group/sub_group overload.
+template <typename GroupHelper, typename InputIteratorT, typename OutputT,
+          std::size_t ElementsPerWorkItem,
+          typename Properties = decltype(properties())>
+std::enable_if_t<
+    std::is_convertible_v<remove_decoration_t<typename std::iterator_traits<
+                              InputIteratorT>::value_type>,
+                          OutputT> &&
+    detail::is_generic_group_v<GroupHelper>>
+group_load(GroupHelper gh, InputIteratorT in_ptr,
+           span<OutputT, ElementsPerWorkItem> out, Properties properties = {}) {
+  constexpr bool blocked = detail::is_blocked(properties);
+
+  using value_type = remove_decoration_t<
+      typename std::iterator_traits<InputIteratorT>::value_type>;
+
+  // See std::enable_if_t above restricting this implementation.
+  using GroupTy = GroupHelper;
+  auto g = gh;
+
+  auto generic = [&]() {
+    group_barrier(g);
+    detail::loop_unroll_up_to<ElementsPerWorkItem, 16>([&](size_t i) {
+      auto idx = detail::get_mem_idx<blocked, ElementsPerWorkItem>(g, i);
+      out[i] = in_ptr[idx];
+    });
+    group_barrier(g);
+  };
+
+  auto delegate = [&](auto unwrapped_ptr) {
+    group_load(gh, unwrapped_ptr, out, properties);
+  };
+
+  constexpr int BlockSize =
+      sizeof(value_type) * (blocked ? ElementsPerWorkItem : 1);
+  constexpr int NumBlocks = blocked ? 1 : ElementsPerWorkItem;
+
+  constexpr auto hw_block_size = [&]() {
+    size_t size = 8;
+    while (BlockSize % size != 0)
+      size /= 2;
+    return size;
+  }();
+  using HWBlockTy = detail::cl_unsigned<hw_block_size>;
+
+  constexpr const size_t hw_blocks_per_block = BlockSize / hw_block_size;
+
+  auto impl_sg = [&](sub_group sg, auto *in_ptr) {
+    value_type v[ElementsPerWorkItem];
+
+    auto priv_ptr = reinterpret_cast<char *>(&v);
```

Comment:
I don't think it matters for this case, but to my knowledge the `&` isn't needed here.